### PR TITLE
Fix instances for plam by making them overlapping and not incoherent

### LINKED ISSUE #2: Improve type inference of plam
In definitions of data types, plam will often fail unless given an explicit type.

See https://github.com/Plutonomicon/plutarch/blob/master/Plutarch/Maybe.hs#L14

### PR DESCRIPTION
**Pull request by L-as**
I think I originally made these instances incoherent because they looked slightly
different, which somehow made OVERLAPPING not work. Then I changed them, without noticing
that I could make them coherent.

### CODE REVIEW

**Comment by L-as** [comment]
And unfortunately this makes some other code break, but it seems worth it.

**Comment by L-as** [comment]
This basically breaks the `id` function `plam $ \x -> x`...

**Comment by L-as** [comment]
#150

**Comment by L-as** [comment]
We probably want to merge this regardless of #150, since it makes error messages much more comprehendable when it does fail:
```
examples/Main.hs:199:37: error:
    • Overlapping instances for Plutarch.PLamN
                                  (p0 -> p0) (Term s1 (PInteger @{S} :--> a0))
        arising from a use of ‘plam’
      Matching instances:
        instance forall a' (s :: S) (a :: PType) b' (b :: PType).
                 ((a' :: Type) ~ (Term s a :: Type),
                  (b' :: Type) ~ (Term s b :: Type)) =>
                 Plutarch.PLamN (a' -> b') (Term s (a :--> b))
          -- Defined in ‘Plutarch’
        instance [overlapping] forall a' (s :: S) (a :: PType) b'
                                      (b :: PType) c' (c :: PType) d' (d :: PType) e' (e :: PType)
                                      f' (f :: PType) g' (g :: PType).
                               ((a' :: Type) ~ (Term s a :: Type),
                                (b' :: Type) ~ (Term s b :: Type),
                                (c' :: Type) ~ (Term s c :: Type),
                                (d' :: Type) ~ (Term s d :: Type),
                                (e' :: Type) ~ (Term s e :: Type),
                                (f' :: Type) ~ (Term s f :: Type),
                                (g' :: Type) ~ (Term s g :: Type)) =>
                               Plutarch.PLamN
                                 (a' -> b' -> c' -> d' -> e' -> f' -> g')
                                 (Term s (a :--> (b :--> (c :--> (d :--> (e :--> (f :--> g)))))))
          -- Defined in ‘Plutarch’
        instance [overlapping] forall a' (s :: S) (a :: PType) b'
                                      (b :: PType) c' (c :: PType) d' (d :: PType) e' (e :: PType)
                                      f' (f :: PType).
                               ((a' :: Type) ~ (Term s a :: Type),
                                (b' :: Type) ~ (Term s b :: Type),
                                (c' :: Type) ~ (Term s c :: Type),
                                (d' :: Type) ~ (Term s d :: Type),
                                (e' :: Type) ~ (Term s e :: Type),
                                (f' :: Type) ~ (Term s f :: Type)) =>
                               Plutarch.PLamN
                                 (a' -> b' -> c' -> d' -> e' -> f')
                                 (Term s (a :--> (b :--> (c :--> (d :--> (e :--> f))))))
          -- Defined in ‘Plutarch’
        ...plus three others
        (use -fprint-potential-instances to see them all)
      (The choice depends on the instantiation of ‘p0, s1, a0’
       To pick the first instance above, use IncoherentInstances
       when compiling the other instance declarations)
    • In the first argument of ‘($)’, namely ‘plam’
      In the second argument of ‘($)’, namely ‘plam $ \ x -> x’
      In the first argument of ‘(#)’, namely
        ‘(phoistAcyclic $ plam $ \ x -> x)’
    |
199 |         printTerm ((phoistAcyclic $ plam $ \x -> x) # (0 :: Term s PInteger)) @?= "(program 1.0.0 0)"
```

**Comment by L-as** [comment]
I made instances incoherent again, but I had to add the pragma to the module rather than to the instances, because they need to be both overlapping and incoherent. Is there a way of doing this?

**Comment by L-as** [comment]
It still fails when doing things like `plam (-)`, but it's good enough in 99% of cases AFAICT.

**Comment by L-as** [comment]
This doesn't work well at all actually, but semiworks within GHCi when you add the language pragma after loading the project...

File: examples/Examples/List.hs [hunk 1]
```diff
@@ -55,14 +55,15 @@ tests = do
           (pzipWith' (+) # integerList [1 .. 10] # integerList [1 .. 10])
             #== integerList (fmap (* 2) [1 .. 10])
     , testCase "pfoldl" $ do
+        let m :: Term _ (PInteger :--> PInteger :--> PInteger) = plam $ \x y -> x - y
         expect $
-          (pfoldl # plam (-) # 0 # integerList [1 .. 10])
+          (pfoldl # m # 0 # integerList [1 .. 10])
```
Comment (blamario) [inline examples/Examples/List.hs:60]: What's the reason for this change?

**Review by blamario — APPROVED** [review]

**Comment by TotallyNotChase** [comment]
Is this contending against #150 ? As in, I assume you're planning to merge only one of these?

**Comment by L-as** [comment]
They don't conflict, but merging #150 would entail eventually removing the curried syntax, probably in Plutarch 1.2, and deprecating them for 1.1.

File: examples/Examples/List.hs (hunk 1 above)
Comment (L-as) [reply examples/Examples/List.hs:60]: Making them coherent breaks this. In general the curried syntax just doesn't work well.

Comment (L-as) [reply examples/Examples/List.hs:60]: The type of `m` also needs to be written explicitly unfortunately.

**Pull request merged**